File: haskell/glue/src/Glue/Lib/List/Position.hs
module Glue.Lib.List.Position where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception
import Glue.IR (IR (..))

position :: [IR Eval] -> Eval (IR Eval)
position [predicateIR, listIR] = do
    predicate <- eval predicateIR
    list <- eval listIR
    case list of
        List xs -> do
            -- Find index of first element that satisfies predicate
            findPosition predicate xs 0
        _ -> throwError $ wrongArgumentType ["function", "list"]
position _ = throwError wrongNumberOfArguments

-- Helper function to find position of first element satisfying predicate
findPosition :: IR Eval -> [IR Eval] -> Int -> Eval (IR Eval)
findPosition _ [] _ = throwError $ wrongArgumentType ["element satisfying predicate"]
findPosition predicate (x : xs) idx = do
    -- Evaluate (predicate x) and check if it returns true
    result <- eval (List [predicate, x])
    case result of
        Bool True -> pure $ Integer (fromIntegral idx)
        Bool False -> findPosition predicate xs (idx + 1)
        _ -> throwError $ wrongArgumentType ["boolean result from predicate"]

File: dart/glue/lib/src/lib/list/position.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Position function - finds index of first element that satisfies a predicate
/// Mirrors Haskell Glue.Lib.List.Position.position exactly
Ir position = IrNativeFunc(positionImpl);

/// Position function implementation
/// Mirrors Haskell Glue.Lib.List.Position.positionImpl exactly
Eval<Ir> positionImpl(Ir predicateIr) {
  return Eval.pure(IrNativeFunc(positionIn(predicateIr)));
}

/// Helper function for list argument
/// Mirrors Haskell Glue.Lib.List.Position.positionIn exactly
Eval<Ir> Function(Ir) positionIn(Ir predicate) {
  return (Ir list) {
    return switch (list) {
      IrList(elements: final elements) => findPosition(
        predicate,
        elements.toList(),
        0,
      ),
      _ => throwError(wrongArgumentType(['function', 'list'])),
    };
  };
}

/// Helper function to find position of first element satisfying predicate
Eval<Ir> findPosition(Ir predicate, List<Ir> elements, int index) {
  if (elements.isEmpty) {
    return throwError(wrongArgumentType(['element satisfying predicate']));
  }

  return applyPredicate(predicate, elements[0]).bind((satisfies) {
    if (satisfies) {
      return Eval.pure(IrInteger(index));
    } else {
      return findPosition(predicate, elements.sublist(1), index + 1);
    }
  });
}

/// Helper function to apply predicate to an element
Eval<bool> applyPredicate(Ir predicate, Ir element) {
  return eval(IrList([predicate, element])).bind((result) {
    if (result is IrBool) {
      return Eval.pure(result.value);
    } else {
      return throwError(wrongArgumentType(['boolean result from predicate']));
    }
  });
}
